### PR TITLE
Accidents and some refinements on speed90

### PR DESCRIPTION
**Pull request by ettorepuccetti**
now fully works with parallelism, but waiting for testing with the new CarEvent that extends Tuple8

### CODE REVIEW

File: flink_project/src/main/java/master2018/flink/accident/FilterZero.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package master2018.flink.accident;
+
+import master2018.flink.data.CarEvent;
+import org.apache.flink.api.common.functions.FilterFunction;
+
+
+public class FilterZero implements FilterFunction<CarEvent> {
+    @Override
+    public boolean filter(CarEvent carEvent) {
+        if(carEvent.getSpeed() == 0) return true;
```
Comment (borismattijssen) [inline flink_project/src/main/java/master2018/flink/accident/FilterZero.java]:
```suggestion
        return (carEvent.getSpeed() == 0);
```

File: flink_project/src/main/java/master2018/flink/accident/FilterZero.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+        else return false;
```
Comment (borismattijssen) [inline flink_project/src/main/java/master2018/flink/accident/FilterZero.java]:
```suggestion
```

File: flink_project/src/main/java/master2018/flink/accident/CountFour.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import master2018.flink.data.CarEvent;
+import master2018.flink.data.Direction;
+
+public class CountFour implements WindowFunction<CarEvent, AccidentEvent, Tuple, GlobalWindow> {
+    
+    @Override
+    public void apply(Tuple key, GlobalWindow window, Iterable<CarEvent> input, Collector<AccidentEvent> out) throws Exception {
+
+        Iterator<CarEvent> iterator = input.iterator();
+        List<CarEvent> listevent = new ArrayList<CarEvent>();
```
Comment (borismattijssen) [inline flink_project/src/main/java/master2018/flink/accident/CountFour.java]:
```suggestion
        List<CarEvent> listevent = new ArrayList<>();
```

File: flink_project/src/main/java/master2018/flink/accident/CountFour.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+import org.apache.flink.streaming.api.windowing.windows.TimeWindow;
+import org.apache.flink.util.Collector;
+
+import master2018.flink.data.CarEvent;
+import master2018.flink.data.Direction;
+
+public class CountFour implements WindowFunction<CarEvent, AccidentEvent, Tuple, GlobalWindow> {
+    
+    @Override
+    public void apply(Tuple key, GlobalWindow window, Iterable<CarEvent> input, Collector<AccidentEvent> out) throws Exception {
```
Comment (borismattijssen) [inline flink_project/src/main/java/master2018/flink/accident/CountFour.java]:
```suggestion
    public void apply(Tuple key, GlobalWindow window, Iterable<CarEvent> input, Collector<AccidentEvent> out) {
```

File: flink_project/src/main/java/master2018/flink/accident/CountFour.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+        Iterator<CarEvent> iterator = input.iterator();
+        List<CarEvent> listevent = new ArrayList<CarEvent>();
+
+        iterator.forEachRemaining(listevent::add);
```
Comment (borismattijssen) [inline flink_project/src/main/java/master2018/flink/accident/CountFour.java]: Fancy spancy 🎉

File: flink_project/src/main/java/master2018/flink/accident/CountFour.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        
+        if (listevent.size() == 4) {
+            CarEvent first = listevent.get(0);
+            CarEvent last = listevent.get(3);
+            boolean timecorrect = true;
+            int timestart = first.getTime();
+            int i = 1;
+            while (timecorrect && i < 4) {
```
Comment (borismattijssen) [inline flink_project/src/main/java/master2018/flink/accident/CountFour.java]: Why would time not be correct? Isn't this weird to check?

File: flink_project/src/main/java/master2018/flink/accident/MyTimeAssigner.java [hunk 7]
```diff
@@ -0,0 +1,13 @@
+package master2018.flink.accident;
+
+import org.apache.flink.streaming.api.functions.timestamps.AscendingTimestampExtractor;
+
+import master2018.flink.data.CarEvent;
+
+public class MyTimeAssigner extends AscendingTimestampExtractor<CarEvent> {
```
Comment (borismattijssen) [inline flink_project/src/main/java/master2018/flink/accident/MyTimeAssigner.java]: This class is never used, so you can throw it away!

**Review by borismattijssen — COMMENTED** [review]
This looks all really good! I left you some small programming tips and I raised one question. 
Also good work on using the whole pull request set-up. I hope you liked it.

Last question: How do verify that your results are good? Do you still use that python tool from your friend?

**Comment by ettorepuccetti** [comment]
Hi,

Thank you for your suggestions, just some questions about changes that I don’t understand:

________________________________

In flink_project/src/main/java/master2018/flink/accident/CountFour.java<https://github.com/ettorepuccetti/flink/pull/3#discussion_r231700575>:

> +import org.apache.flink.api.java.tuple.Tuple;
+import org.apache.flink.streaming.api.functions.windowing.WindowFunction;
+import org.apache.flink.streaming.api.windowing.windows.GlobalWindow;
+import org.apache.flink.streaming.api.windowing.windows.TimeWindow;
+import org.apache.flink.util.Collector;
+
+import master2018.flink.data.CarEvent;
+import master2018.flink.data.Direction;
+
+public class CountFour implements WindowFunction<CarEvent, AccidentEvent, Tuple, GlobalWindow> {
+
+    @Override
+    public void apply(Tuple key, GlobalWindow window, Iterable<CarEvent> input, Collector<AccidentEvent> out) throws Exception {
+
+        Iterator<CarEvent> iterator = input.iterator();
+        List<CarEvent> listevent = new ArrayList<CarEvent>();


⬇️ Suggested change

-        List<CarEvent> listevent = new ArrayList<CarEvent>();
+        List<CarEvent> listevent = new ArrayList<>();


Why do not I have to specify the type of ArrayList?


________________________________

In flink_project/src/main/java/master2018/flink/accident/CountFour.java<https://github.com/ettorepuccetti/flink/pull/3#discussion_r231701944>:

> +
+    @Override
+    public void apply(Tuple key, GlobalWindow window, Iterable<CarEvent> input, Collector<AccidentEvent> out) throws Exception {
+
+        Iterator<CarEvent> iterator = input.iterator();
+        List<CarEvent> listevent = new ArrayList<CarEvent>();
+
+        iterator.forEachRemaining(listevent::add);
+
+        if (listevent.size() == 4) {
+            CarEvent first = listevent.get(0);
+            CarEvent last = listevent.get(3);
+            boolean timecorrect = true;
+            int timestart = first.getTime();
+            int i = 1;
+            while (timecorrect && i < 4) {


Why would time not be correct? Isn't this weird to check?

In the Count windows I”m receiving 4 events from the same car, in the same position. But there are no guarantees that they are consecutive (every 30 seconds) so I have to check it.
Also I rename a couple of variables and class name to make it easier to understand.


I just have a look at the two new branches you made on the repo, how should I proceed for merging all the work? It look like a bit confusing…

Ettore

**Pull request merged**